engine/common/imagelib/img_ktx2.c

```diff
@@ -0,0 +1,204 @@
+	if( header->supercompressionScheme != 0 )
+	{
+		Con_DPrintf( S_ERROR "%s: unsupported KTX2 supercompressionScheme %d\n", __FUNCTION__, header->supercompressionScheme );
+		return false;
+	}
+
+	if( header->levelCount * sizeof( ktx2_level_t ) + KTX2_LEVELS_OFFSET > filesize )
+	{
+		Con_DPrintf( S_ERROR "%s: file abruptly ends\n", __FUNCTION__ );
+		return false;
+	}
+
+	memcpy( &index, buffer + KTX2_IDENTIFIER_SIZE + sizeof( ktx2_header_t ), sizeof( index ));
+
+	for( int mip = 0; mip < header->levelCount; ++mip )
+	{
+		const uint32_t width = Q_max( 1, ( header->pixelWidth >> mip ));
+		const uint32_t height = Q_max( 1, ( header->pixelHeight >> mip ));
+		const uint32_t mip_size = Image_ComputeSize( image.type, width, height, image.depth );
+
+		ktx2_level_t level;
+		memcpy( &level, levels_begin + mip * sizeof( level ), sizeof( level ));
+
+		if( mip_size != level.byteLength )
+		{
+			Con_DPrintf( S_ERROR "%s: mip=%d size mismatch read=%d, but computed=%d\n",
+				__FUNCTION__, mip, (int)level.byteLength, mip_size );
+			return false;
+		}
+
+		total_size += level.byteLength;
+		max_offset = Q_max( max_offset, level.byteLength + level.byteOffset );
+	}
+
+	if( max_offset > filesize )
+		return false;
+
+	image.size = total_size;
+	image.num_mips = header->levelCount;
+
+	image.rgba = Mem_Malloc( host.imagepool, image.size );
+	memcpy( image.rgba, buffer, image.size );
+
+	for( int mip = 0, cursor = 0; mip < header->levelCount; ++mip )
+	{
+		ktx2_level_t level;
+		memcpy( &level, levels_begin + mip * sizeof( level ), sizeof( level ));
+		memcpy( image.rgba + cursor, buffer + level.byteOffset, level.byteLength );
+		cursor += level.byteLength;
+	}
+
+	return true;
+}
+
+qboolean Image_LoadKTX2( const char *name, const byte *buffer, fs_offset_t filesize )
+{
+	ktx2_header_t header;
+
+	if( filesize < KTX2_MINIMAL_HEADER_SIZE )
+		return false;
+
+	if( memcmp( buffer, KTX2_IDENTIFIER, KTX2_IDENTIFIER_SIZE ) != 0 )
+	{
+		Con_DPrintf( S_ERROR "%s: (%s) has invalid identifier\n", __FUNCTION__, name );
+		return false;
+	}
+
+	memcpy( &header, buffer + KTX2_IDENTIFIER_SIZE, sizeof( header ));
+
+	image.width = header.pixelWidth;
+	image.height = header.pixelHeight;
+	image.depth = Q_max( 1, header.pixelDepth );
+	image.num_mips = 1;
+
+	ClearBits( image.flags, IMAGE_HAS_COLOR | IMAGE_HAS_ALPHA | IMAGE_HAS_LUMA );
+
+	if( !Image_KTX2Parse( &header, buffer, filesize ))
+	{
+		if( !Image_CheckFlag( IL_KTX2_RAW ))
+			return false;
+
+		// If KTX2 to imagelib conversion failed, try passing the file as raw data.
+		// This is useful for ref_vk which can directly support hundreds of formats which we don't convert to pixformat_t here
+
+		Con_DPrintf( S_WARN "%s: (%s) could not be converted to supported imagelib format, passing as raw KTX2 data\n", __FUNCTION__, name );
+		// This is a catch-all for ref_vk, which can do this format directly and natively
+		image.type = PF_KTX2_RAW;
+
+		image.size = filesize;
+		//image.encode = TODO custom encode type?
```

Comment:
I wonder if `image.encode` would be a good place for format semantic flags like `SRGB`, `UNORM/UNSIGNED`, and `SNORM/UNSIGNED`.

Reply:
Not really sure what should go into `encode` field in this case. I'm fine with extending `pixformat_t` anyway.